src/formatting_stack/util/diff.clj

```diff
@@ -0,0 +1,71 @@
+(ns formatting-stack.util.diff
+  (:require
+   [clojure.java.data :as data]
+   [clojure.java.io :as io]
+   [clojure.spec.alpha :as spec]
+   [formatting-stack.protocols.spec :as protocols.spec]
+   [clojure.string :as string]
+   [formatting-stack.util :refer [rcomp]]
+   [nedap.speced.def :as speced]
+   [nedap.utils.spec.predicates :refer [present-string?]])
+  (:import
+   (difflib DiffUtils)
+   (io.reflectoring.diffparser.api UnifiedDiffParser)
+   (java.io File)
+   (java.util.regex Pattern)))
+
+(spec/def ::start ::protocols.spec/line)
+(spec/def ::end ::protocols.spec/line)
+(spec/def ::filename present-string?)
+
+(speced/def-with-doc ::line-numbers
+  "maps of consecutive removals/changes (as `:start` to `:end`) per `:filename`."
+  (spec/coll-of (spec/keys :req-un [::start ::end ::filename])))
+
+(speced/defn ^::line-numbers diff->line-numbers
+  [^string? diff]
+  (->> (.parse (UnifiedDiffParser.) (io/input-stream (.getBytes diff)))
+       (data/from-java)
+       (mapcat (speced/fn [{:keys [^present-string? toFileName ^coll? hunks]}]
+                 (->> hunks
+                      (mapcat (speced/fn [{:keys [^coll? lines] {:keys [^::start lineStart]} :fromFileRange}]
+                                (->> lines
+                                     (remove (rcomp :lineType #{"TO"}))
+                                     (map-indexed (fn [idx line] (assoc line :lineNumber (+ idx lineStart)))))))
+                      (reduce (speced/fn [ret {:keys [^{::speced/spec #{"FROM" "NEUTRAL"}} lineType
+                                                      ^::protocols.spec/line lineNumber]}]
+                                (let [{:keys [end] :as current} (last ret)]
+                                  (cond
+                                    (#{"NEUTRAL"} lineType)
+                                    ret
+
+                                    (= end (dec lineNumber))
+                                    (assoc ret (dec (count ret)) (assoc current :end lineNumber)) ;; update last map
+
+                                    :else
+                                    (conj ret {:filename (string/replace-first toFileName "b/" "")
+                                               :start lineNumber
+                                               :end   lineNumber}))))
+                              []))))))
+
+(def diff-context-size 3)
+
+(def ^:dynamic *to-absolute-path-fn*
```

Comment:
This could use a small docstring explaining why it is `^:dynamic` and how that is used (figured out this is useful in tests, only by reviewing all of these changes, will be harder once this is actually part of the codebase